### ISSUE TITLE
Password policy enforcement when using Active Directory

### ISSUE THREAD

**Issue by Clay00002** [adldap/adLDAP#63]
I have access to an Active Directory that enforces a 2 password history restriction. Any password in the last 2 passwords you have, is not a viable candidate to be set or reset as your password.

Is there any way have ldap respect this restriction?

**Comment by samwilson** [comment]
Surely adLDAP is not able to circumvent that?! I'm surprised. We've got adLDAP working with AD, and most certainly it respects the password history requirements (although, it can be a pain to report why a password-change has not worked).

**Comment by strebl** [comment]
Don't forget the difference between a password **change** and a password **reset**.
As I know, adldap does not support password changes. And a reset password seems to ignore the password policy.

**Comment by strebl** [comment]
Try to set the same password in the active directory users and computers.
In our environment I can set a password which is in the history but I can't set a to short password.

**Comment by Clay00002** [comment]
My server is window 2003 server, and we enable GPO( password history ) . 
In our test , it's true that we can't handle this rule .
But we can deal with other rules. ( like password length, password complexity )

However, we test it on window 2008R2 and we get the information from microsoft (http://laviefrugale.blogspot.com.br/2013/01/enforcing-active-directory-password.html).
We followed this example and test it. 
Sadly , it wasn't work. 
(We always get this Warning:  Server is unwilling to perform.)

Our experience is like strebl said: 
"reset password seems to ignore the password history policy in some circumstances."

But microsoft sent the information. (That really make we crazy..)
Is there any way have ldap respect this restriction?

**Comment by strebl** [comment]
Yes, if you change the password instead of setting it, you should get an error if you trying to change it with a password in the history.

**Comment by ChadSikorra** [comment]
@Clay00002 As far as I know, you could use `ldap_modify_batch` to set the password in such a way that AD will respect the password requirements. See example number 5 on the docs for the function: http://php.net/manual/en/function.ldap-modify-batch.php . Of course, this is a newer LDAP function and requires a relatively current version of PHP.

**Comment by Clay00002** [comment]
We tried this function before! It is on PHP 5.5 . But windows 2003 is only available for PHP 5.3.

We also try it on Windows server 2012.

In expample 5, It is not working.
We get the Warning: Constraint violation.
 But we change like this:


			$modifs = [
					    [
					        "attrib"  => "unicodePwd",
					        "modtype" => LDAP_MODIFY_BATCH_REPLACE,
					        "values"  => [$sPassword],
					    ],
				 ];

			$Result = ldap_modify_batch($AD_CONNECTION, $sDn, $modifs);

It works !
We can not understand why we get the warning.
And there is not much example on Google or StackOverFolw

**Comment by ChadSikorra** [comment]
@Clay00002 Odd, example 5 for `ldap_modify_batch` works fine in the test domain I have setup. I just tested it actually. Are you using using an encrypted connection (ie. issue a `ldap_start_tls` after you connect to LDAP)? Does binding to LDAP as the user whose password you are going to change make any difference?

**Comment by strebl** [comment]
@Clay00002 That is not a password change anymore, that's a password reset / set.

To change a password the modify request must contain a delete and a add operation.

The replace operations only works if the binding user has sufficient permissions to modify another user's password.

Maybe that helps you: https://support.microsoft.com/en-us/kb/269190

**Comment by Clay00002** [comment]
If we want to change AD password, we must have LDAPS( LDAP over ssl ).
So I think the code is fine!( because I can reset password using by ldap_modify_batch. But not change)

@ChadSikorra Maybe you can tell me what php version you use and what version about your server os. I will try it.


We think the point as @strebl said.
How to "change" AD password but not reset !

However, we tried ldap_mod_delete to delete unicodePwd and it's not working.
we also tried this example:

                   <?php
function adifyPw($pw)
{
    return iconv("UTF-8", "UTF-16LE", '"' . $pw . '"');
}

$dn = "cn=Jack Smith-Jones,ou=Wizards,dc=ad,dc=example,dc=com";
$modifs = [
    [
        "attrib"  => "unicodePwd",
        "modtype" => LDAP_MODIFY_BATCH_REMOVE,
        "values"  => [adifyPw("Tr0ub4dor&3")],
    ],
    [
        "attrib"  => "unicodePwd",
        "modtype" => LDAP_MODIFY_BATCH_ADD,
        "values"  => [adifyPw("correct horse battery staple")],
    ],
];
ldap_modify_batch($connection, $dn, $modifs);

But still not working!( We test all on window server 2008R2 and server 2012 R2.)
Maybe PHP is not working like this. ( But .net can afford this. Maybe we tried to call .net function... )

**Comment by ralfbecker** [comment]
Found why it failed in my test-implementation at the end of following blog post:

   http://blogs.forgerock.org/petermajor/2013/08/fun-with-active-directory/

Seems my used Windows 2008r2 has by default a minimum password-age of one day. Which causes a Constrain Violation, if a regular user trys to change his password more often then once a day.

To remove that restriction you need to run:

   net accounts /MINPWAGE:0

After that code from php.net works for me:

   https://github.com/ralfbecker/adLDAP/commit/21c4680b4393ea35ea82b6d80a40c561ec742606

Ralf

**Comment by Ahmad44** [comment]
Is it possible to enforce password policies while resetting in Active Directory. I know it is an administrative activity but I've just heard a new that we can enforce password policy while resetting.
